paddle/fluid/operators/distributed_ops/recv_save_op.cc

```diff
@@ -0,0 +1,255 @@
+  }
+
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    auto place = ctx.GetPlace();
+
+    auto filename = ctx.Attr<std::string>("file_path");
+    auto overwrite = ctx.Attr<bool>("overwrite");
+
+    if (FileExists(filename) && !overwrite) {
+      PADDLE_THROW("%s is existed, cannot save to it when overwrite=false",
+                   filename, overwrite);
+    }
+
+    MkDirRecursively(DirName(filename).c_str());
+
+    auto origin_shape = ctx.Attr<std::vector<int64_t>>("shape");
+    auto slice_shapes = ctx.Attr<std::vector<std::string>>("slice_shapes");
+    auto slice_varnames = ctx.Attr<std::vector<std::string>>("slice_varnames");
+    auto remote_varnames =
+        ctx.Attr<std::vector<std::string>>("remote_varnames");
+    auto endpoints = ctx.Attr<std::vector<std::string>>("endpoints");
+
+    PADDLE_ENFORCE_EQ(slice_shapes.size(), slice_varnames.size());
+    PADDLE_ENFORCE_EQ(slice_shapes.size(), endpoints.size());
+
+    auto data_type =
+        static_cast<framework::proto::VarType::Type>(ctx.Attr<int>("dtype"));
+
+    // it to save an output stream.
+    std::ofstream fout(filename, std::ios::binary);
+    PADDLE_ENFORCE(static_cast<bool>(fout), "Cannot open %s to write",
+                   filename);
+
+    SerializeVersionToStream(fout);
+    SerializeTensorHeaderToStream(fout, data_type,
+                                  framework::make_ddim(origin_shape));
+
+    // framework::Scope &local_scope = ctx.scope().NewScope();
+    std::unique_ptr<framework::Scope> local_scope = ctx.scope().NewTmpScope();
+
+    auto trainer_id = ctx.Attr<int>("trainer_id");
+
+    platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
+    auto &device_ctx = *pool.Get(place);
+
+    distributed::RPCClient *rpc_client =
+        distributed::RPCClient::GetInstance<RPCCLIENT_T>(trainer_id);
+
+    for (int i = 0; i < slice_varnames.size(); i++) {
+      auto &varname = slice_varnames[i];
+      auto *var = local_scope->Var(varname);
+      auto *tensor = var->GetMutable<framework::LoDTensor>();
+
+      auto slice_string =
+          string::split_string<std::string>(slice_shapes[i], ",");
+      std::vector<int64_t> slice_shape;
+
+      for (auto &dim : slice_string) {
+        slice_shape.push_back(static_cast<int64_t>(std::stoull(dim)));
+      }
+
+      tensor->Resize(framework::make_ddim(slice_shape));
+
+      distributed::VarHandlePtr ret;
+
+      ret = rpc_client->AsyncGetVarNoBarrier(endpoints[i], device_ctx,
+                                             *local_scope.get(),
+                                             remote_varnames[i], varname);
+
+      PADDLE_ENFORCE_NE(ret->Wait(), 0U, "internal error in RPCClient");
+
+      auto &c_tensor = var->Get<framework::LoDTensor>();
+
+      SerializeTensorAppendToStream(fout, c_tensor);
+      local_scope->EraseVars({varname});
+    }
+    fout.close();
+    local_scope.release();
```

Comment:
I think this can not delete framework::Scope pointer. release() only gives up the control of the Scope pointer.

Reply:
done